dhib_env.py

```diff
@@ -0,0 +1,456 @@
+#!/usr/bin/env python3
+
+""" A script to build a virtual environment for Deephaven-IB development or release."""
+
+import atexit
+import logging
+import os
+import re
+import shutil
+from pathlib import Path
+from types import ModuleType
+from typing import Optional, Dict, Union
+import click
+import pkginfo
+import requests
+
+IB_VERSION_DEFAULT="10.19.04"
+DH_VERSION_DEFAULT="0.33.0"
+
+########################################################################################################################
+# Version Numbers
+########################################################################################################################
+
+
+def version_tuple(version: str) -> tuple[int, ...]:
+    """Convert a version string to a tuple of integers.
+
+    Args:
+        version: The version string to convert.
+
+    Returns:
+        A tuple of integers representing the version.
+    """
+    return tuple(map(int, (version.split("."))))
+
+
+def version_str(version: tuple[int, ...], wide: bool) -> str:
+    """Convert a version tuple to a string.
+
+    Args:
+        version: The version tuple to convert.
+        wide: Whether to use a wide format that includes leading zeros.
+
+    Returns:
+        A string representing the version.
+    """
+    if wide:
+        return ".".join(f"{x:02d}" for x in version)
+    else:
+        return ".".join(map(str, version))
+
+
+def version_assert_format(version: str) -> None:
+    """Assert that a version string is formatted correctly.
+
+    Args:
+        version: The version string to check.
+
+    Raises:
+        ValueError: If the version string is not formatted correctly.
+    """
+    if not version:
+        raise ValueError("Version string is empty.")
+
+    # check if the version string is in semver format
+    pattern1 = re.compile(r"^([0-9]\d*)\.([0-9]\d*)\.([0-9]\d*)$")
+    pattern2 = re.compile(r"^([0-9]\d*)\.([0-9]\d*)\.([0-9]\d*)\.dev([0-9]\d*)$")
+    is_semver = bool(pattern1.match(version)) or bool(pattern2.match(version))
+
+    if not is_semver:
+        raise ValueError(f"Version string is not in semver format: {version}")
+
+
+########################################################################################################################
+# Shell
+########################################################################################################################
+
+
+def shell_exec(cmd: str) -> None:
+    """Execute a shell command.
+
+    Args:
+        cmd: The command to execute.
+    """
+    logging.warning(f"Executing shell command: {cmd}")
+    e = os.system(cmd)
```

Comment:
Any reason to not use `subprocess` to capture stdout/stderr in case the command fails?

Reply:
Yes.  It gave me all kinds of awful problems when I tried to use it.

Reply:
Also, what is here does log stdout/stderr, so the user gets plenty of log spam.